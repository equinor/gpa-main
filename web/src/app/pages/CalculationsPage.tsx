import { useQuery } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import { History } from 'history';
import moment from 'moment';
import React from 'react';
import styled from 'styled-components';

import { IShip } from '../common/Interfaces';
import { CalculationsContainer } from '../components/content/calculations/CalculationsContainer';
import { IStandardTableRow } from '../components/elements/Tables';
import { P } from '../components/elements/Texts';
import { UserMessage } from '../components/elements/UserMessage';
import { PageContent } from '../components/ui/PageContent';
import { TitleBlock } from '../components/ui/TitleBlock';
import { StPageWrapper } from './CalculatorPage';

export interface ICalculationInfo {
  id: string,
  ship: IShip,
  createdDate: string
}

interface ICalculationsData {
  calculations: ICalculationInfo[],
}

interface ICalculationPage {
  history: History;
}

export const CalculationsPage: React.FunctionComponent<ICalculationPage> = ({ history }) => {
  const calculations = useQuery<ICalculationsData>(CALCULATIONS_QUERY);
  let rows: IStandardTableRow[] = [];

  if (calculations.data) {
    rows = calculations.data.calculations.map((calculation: ICalculationInfo) => {
      return {
        value: calculation.id,
        display: [calculation.ship.name, calculation.ship.country, calculation.createdDate ? moment(calculation.createdDate).fromNow() : ""] 
      }
    })
  }

  return (
    <PageContent>
      <TitleBlock>Results</TitleBlock>
      <StPageWrapper>
        <P>Select a row from the table to see details of the calculation.</P>
        <div style={{ margin: "30px 0 0 0" }}>
          {calculations.loading &&
            <StUserMessage>
              <UserMessage type={"loading"} text={"Loading results"} />
            </StUserMessage> 
          } 
          {calculations.data &&
            <CalculationsContainer
              rows={rows}
              selectRow={(row: IStandardTableRow) => {
                history.push('/result/' + row.value);
              }}
            ></CalculationsContainer> 
          }
          {calculations.error &&
            <StUserMessage>
              <UserMessage type={"error"} text={"Loading failed"} />
            </StUserMessage>
          }
        </div>
      </StPageWrapper>
    </PageContent >
  );
};

const StUserMessage = styled.div`
    display: flex;
    justify-content: center;
    margin: 50px 0 0;
`;

export const CALCULATIONS_QUERY = gql`
    {
        calculations {
            id
            ship {
                id
                name
                country
            }
            createdDate
        }
    }
`;
