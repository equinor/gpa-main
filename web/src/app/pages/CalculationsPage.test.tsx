//TESTING DATA, LOADING, ERROR snapshots
import { MockedProvider, wait } from '@apollo/react-testing';
import { createMemoryHistory } from 'history';
import React from 'react';
import renderer, { act } from 'react-test-renderer';

import {CalculationsPage, CALCULATIONS_QUERY } from './CalculationsPage';

describe("CalculationSPage", () => {
    //loading
    it("snapshots to loading", () => {
        const history = createMemoryHistory(undefined);
        //loading
        const tree = renderer.create(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CalculationsPage history={history}></CalculationsPage>
            </MockedProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })

    //data
    it("snapshots to data", async () => {
        const history = createMemoryHistory(undefined);
        //loading
        const component = renderer.create(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CalculationsPage history={history}></CalculationsPage>
            </MockedProvider>
        )

        await act(async () => {
            await wait(0);
            expect(component.toJSON()).toMatchSnapshot();
        });
    }) 

    //error 
    it("snapshots to error", async () => {
        const history = createMemoryHistory(undefined);
        //loading
        const component = renderer.create(
            <MockedProvider mocks={mocksError} addTypename={false}>
                <CalculationsPage history={history}></CalculationsPage>
            </MockedProvider>
        )

        await act(async () => {
            await wait(0);
            expect(component.toJSON()).toMatchSnapshot();
        });
    })
})   

const mocks = [
    {
        request: {
            query: CALCULATIONS_QUERY
        },
        result: {
            data: {
                calculations:
                    [
                        {
                            "id": "103",
                            "ship": {
                                "id": "104",
                                "name": "a",
                                "country": "b",
                                "__typename": "Ship"
                            },
                            "createdDate": null,
                            "__typename": "Calculation"
                        },
                        {
                            "id": "1",
                            "ship": {
                                "id": "2",
                                "name": "ShipName",
                                "country": "ShipCountry",
                                "__typename": "Ship"
                            },
                            "createdDate": null,
                            "__typename": "Calculation"
                        }
                    ]
            }
        }
    }
]

const mocksError = [{
    request: {
        query: CALCULATIONS_QUERY
    },
    error: new Error()
}]