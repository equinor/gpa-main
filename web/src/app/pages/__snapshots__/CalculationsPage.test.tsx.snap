// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculationSPage snapshots to data 1`] = `
<div
  className="PageContent__StPageContent-sc-1m5vq8g-0 kNVjng"
>
  <div
    className="TitleBlock__StTitleBlock-sc-1jpbjrp-0 cGhYnk"
  >
    <h1>
      Results
    </h1>
  </div>
  <div
    className="CalculatorPage__StPageWrapper-h4fxsn-0 gfXgRu"
  >
    <p
      className="Texts__P-teybz3-2 grKhWK"
    >
      Select a row from the table to see details of the calculation.
    </p>
    <div
      style={
        Object {
          "margin": "30px 0 0 0",
        }
      }
    >
      <div>
        <table
          className="Tables__StStandardTable-sc-5p5bny-1 jmmnft"
        >
          <thead>
            <tr>
              <td>
                NAME
              </td>
              <td>
                COUNTRY
              </td>
              <td>
                CREATED
              </td>
            </tr>
          </thead>
          <tbody>
            <tr
              onClick={[Function]}
            >
              <td>
                a
              </td>
              <td>
                b
              </td>
              <td>
                
              </td>
            </tr>
            <tr
              onClick={[Function]}
            >
              <td>
                ShipName
              </td>
              <td>
                ShipCountry
              </td>
              <td>
                
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`CalculationSPage snapshots to error 1`] = `
<div
  className="PageContent__StPageContent-sc-1m5vq8g-0 kNVjng"
>
  <div
    className="TitleBlock__StTitleBlock-sc-1jpbjrp-0 cGhYnk"
  >
    <h1>
      Results
    </h1>
  </div>
  <div
    className="CalculatorPage__StPageWrapper-h4fxsn-0 gfXgRu"
  >
    <p
      className="Texts__P-teybz3-2 grKhWK"
    >
      Select a row from the table to see details of the calculation.
    </p>
    <div
      style={
        Object {
          "margin": "30px 0 0 0",
        }
      }
    >
      <div
        className="sc-bZQynM iBlShU"
      >
        <div
          className="sc-bdVaJa cRNkdB"
          type="error"
        >
          <svg
            fill="none"
            height={24}
            viewBox="0 0 10 10"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.66659 1.27334L8.72659 0.333344L4.99992 4.06001L1.27325 0.333344L0.333252 1.27334L4.05992 5.00001L0.333252 8.72668L1.27325 9.66668L4.99992 5.94001L8.72659 9.66668L9.66659 8.72668L5.93992 5.00001L9.66659 1.27334Z"
              fill="#EB0000"
            />
          </svg>
          <span>
            Loading failed
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CalculationSPage snapshots to loading 1`] = `
<div
  className="PageContent__StPageContent-sc-1m5vq8g-0 kNVjng"
>
  <div
    className="TitleBlock__StTitleBlock-sc-1jpbjrp-0 cGhYnk"
  >
    <h1>
      Results
    </h1>
  </div>
  <div
    className="CalculatorPage__StPageWrapper-h4fxsn-0 gfXgRu"
  >
    <p
      className="Texts__P-teybz3-2 grKhWK"
    >
      Select a row from the table to see details of the calculation.
    </p>
    <div
      style={
        Object {
          "margin": "30px 0 0 0",
        }
      }
    >
      <div
        className="sc-bZQynM iBlShU"
      >
        <div
          className="sc-bdVaJa cqSazf"
          type="loading"
        >
          <svg
            fill="none"
            height={24}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M12 1.84615C6.39219 1.84615 1.84615 6.39219 1.84615 12C1.84615 17.6078 6.39219 22.1538 12 22.1538C17.6078 22.1538 22.1538 17.6078 22.1538 12C22.1538 6.39219 17.6078 1.84615 12 1.84615ZM0 12C0 5.37258 5.37258 0 12 0C18.6274 0 24 5.37258 24 12C24 18.6274 18.6274 24 12 24C5.37258 24 0 18.6274 0 12Z"
              fill="#DEEDEE"
              fillRule="evenodd"
            />
            <path
              clipRule="evenodd"
              d="M3.26622 19.1615C3.62672 18.79 4.21764 18.7835 4.58608 19.1471C5.77978 20.3249 7.25115 21.1956 8.87067 21.6804C10.4902 22.1651 12.2065 22.2484 13.8669 21.9225C15.5272 21.5966 17.0788 20.8721 18.3845 19.8147C19.6902 18.7575 20.7095 17.4005 21.354 15.8651C21.9985 14.3299 22.249 12.6625 22.0841 11.0103C21.9191 9.35803 21.3437 7.76994 20.407 6.38681C19.4703 5.00356 18.2008 3.86748 16.7095 3.08099C15.2181 2.29444 13.5519 1.88235 11.8586 1.88235C11.3432 1.88235 10.9253 1.46097 10.9253 0.941176C10.9253 0.421379 11.3432 -4.54098e-07 11.8586 -4.76629e-07C13.8517 -5.63749e-07 15.8149 0.484934 17.5746 1.41299C19.3344 2.34109 20.8372 3.68426 21.9485 5.32528C23.0599 6.96643 23.7449 8.8543 23.9413 10.8218C24.1377 12.7893 23.8392 14.7738 23.073 16.5989C22.3068 18.424 21.0972 20.0322 19.5533 21.2824C18.0095 22.5326 16.1787 23.3864 14.2235 23.7702C12.2684 24.1539 10.2476 24.056 8.33952 23.4849C6.43142 22.9138 4.69353 21.8866 3.28059 20.4925C2.91215 20.129 2.90572 19.5331 3.26622 19.1615Z"
              fill="#007079"
              fillRule="evenodd"
            />
          </svg>
          <span>
            Loading results
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
